File: 2982/main.go
package main

import "fmt"

func main() {
	fmt.Println(maximumLength("abcaba"))
}

func maximumLength(s string) int {
	n := len(s)
	cnt := make(map[byte][]int)
	for i, j := 0, 0; i < n; i = j {
		for j < n && s[j] == s[i] {
			j++
		}
		cnt[s[i]] = append(cnt[s[i]], j-i)
	}
	res := -1
	for _, vec := range cnt {
		lo, hi := 1, n - 2
		for lo <= hi {
			mid := (lo + hi) >> 1
			count := 0
			for _, x := range vec {
				if x >= mid {
					count += x - mid + 1
				}
			}
			if count >= 3 {
				if mid > res {
					res = mid
				}
				lo = mid + 1
			} else {
				hi = mid - 1
			}
		}
	}
	return res
}

File: 2982/main.rs
use std::collections::HashMap;

struct Solution{}

impl Solution {
    pub fn maximum_length(s: String) -> i32 {
        let n = s.len();
        let mut cnt = HashMap::new();
        let s_bytes = s.as_bytes();

        let mut i = 0;
        while i < n {
            let mut j = i;
            while j < n && s_bytes[j] == s_bytes[i] {
                j += 1;
            }
            cnt.entry(s_bytes[i])
                .or_insert_with(Vec::new)
                .push((j-i) as i32);
            i = j;
        }
        let mut res = -1;
        for vec in cnt.values() {
            let mut lo = 1;
            let mut hi = (n as i32) - 2;
            while lo <= hi {
                let mid = (lo + hi) / 2;
                let mut count = 0;
                for &x in vec {
                    if x >= mid {
                        count += x - mid + 1;
                    }
                }
                if count >= 3 {
                    res = res.max(mid);
                    lo = mid + 1;
                } else {
                    hi = mid - 1;
                }
            }
        }
        res
    }
}

fn main() {
    println!("{}", Solution::maximum_length("abcaba".to_string()))
}